treewide: `"${pkg}/bin/${pkg.meta.mainProgram}"` -> `"${lib.getExe pkg}"`

pbsds opened the pull request:
## Description of changes

Made with

```shell
if test ! -f packages.json; then
  nix-env --extra-experimental-features no-url-literals --option system x86_64-linux -f ./. -qaP --json --meta --show-trace --no-allow-import-from-derivation > packages.json
fi

jq <packages.json '
    to_entries[]
    | {attr: .key, exe: .value.meta.mainProgram}
    | select(.exe == null|not)
' -c |
while read line; do
    attr="$(jq <<<"$line" .attr -r)"
    exe="$(jq <<<"$line" .exe -r)"
    echo| ( set -x
        rg -F "\"\${$attr}/bin/$exe\"" -tnix -l pkgs/ | xe sd -F "\"\${$attr}/bin/$exe\"" "\"\${lib.getExe $attr}\""
        rg -F "\"\${lib.getBin $attr}/bin/$exe\"" -tnix -l pkgs/ | xe sd -F "\"\${lib.getBin $attr}/bin/$exe\"" "\"\${lib.getExe $attr}\""
    )
done
```

and reviewed with `git add --patch`.

Should cause zero rebuilds

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

pbsds committed:
treewide: `"${pkg}/bin/${pkg.meta.mainProgram}"` -> `"${lib.getExe pkg}"`

drupol reviewed (approved):
LGTM

AndersonTorres commented on pkgs/applications/audio/aaxtomp3/default.nix:
```diff
@@ -45,7 +45,7 @@ resholve.mkDerivation rec {
       "bin/interactiveaaxtomp3"
     ];
-    interpreter = "${bash}/bin/bash";
+    interpreter = "${lib.getExe bash}";
```
An LSP mage would be cool now!

Or a mere regex-replace...

SuperSandro2000 replied:
Just for context: I fully accept a "lets do that in another PR after this one" I just wanted to bring the idea up and talk about it for a second.

eclairevoyant replied:
I'm leaning towards new PR, since we already have some instances of `"${...}"` in nixpkgs that can be cleaned up at the same time, and we should be careful to only change those examples outside of an indented strinpg where we don't rely implicitly on the path -> string coercion.

Besides, I think this PR is quite large to review as is

pbsds replied:
I did a `rg ' = "\$\{lib.getExe .*\}";' pkgs/ -l | xe sd ' = "\$\{lib.getExe (.*)\}";' ' = lib.getExe $1;'`

afh reviewed (approved):
Excellent work, @pbsds! :ok_hand:

pbsds committed:
treewide: `= "${lib.getExe x}";` -> `= lib.getExe x;`

command used:

    rg ' = "\$\{lib.getExe .*\}";' pkgs/ -l | xe sd ' = "\$\{lib.getExe (.*)\}";' ' = lib.getExe $1;'
